### PR TITLE
Add command synchronization

### LINKED ISSUE #15: Sync Discord Commands On-Demand
Basically don't _always_ sync application commands whenever the bot starts up. Instead the client should have a method `syncCommands()` which synchronizes the commands to the server. In addition a CLI argument can be added: `--sync` which tells the bot to sync with discord servers.

### PR DESCRIPTION
**Pull request by suneettipirneni**
This removes the old behavior of pushing commands to the API on every startup, and only pushes commands whenever a change is detected.

closes #16

### CODE REVIEW

**Comment by rob-3** [comment]
Oh no I think it is good for production; I just change it locally for hacking around.

**Comment by rob-3** [comment]
I think you should remove the eslint suppressions in `Client.ts`.

**Comment by rob-3** [comment]
You can remove the other one too if you add an if to check for the guild id.

**Comment by rob-3** [comment]
I'm starting to think we should store the guild id in a field on the client using the constructor so we only have to check it exists once.

**Comment by suneettipirneni** [comment]
> I'm starting to think we should store the guild id in a field on the client using the constructor so we only have to check it exists once.

A singular guild ID is a temporary thing, we'll remove them later, and have each command have it's own `guildIDs` field, since a bot can be part of multiple guilds.

File: src/utils/command.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  if (!a || !b) {
+    return false;
+  }
+
+  return a.name === b.name &&
+    isEqual(a.options, b.options) &&
+    a.description === b.description;
+}
```
Comment (rob-3) [inline src/utils/command.ts]: Any reason we can't just replace this entire function with `isEqual(a, b)`?

File: src/Client.ts [hunk 2]
```diff
@@ -11,18 +12,82 @@ import discord, {
...
+      await push();
+      return;
+    }
+
+
+    // Calculate if theres a diff between the local and remote commands.
+    const diff = !appCommands.every(appCommand => {
+      // Get the name, and get the corresponding command with the same name.
+      const { name } = appCommand;
+      const clientCommand = clientCommands.get(name);
```
Comment (rob-3) [inline src/Client.ts]: I think you should just replace these lines with `const clientCommand = clientCommands.get(appCommand.name);`

**Comment by rob-3** [comment]
Not done with this review but those are some initial thoughts

**Comment by rob-3** [comment]
One small issue that I see is that this PR doesn't seem to push on a per command basis. Either it detects no changes, so nothing gets pushed, or it detects a change, and then everything gets pushed. Not necessarily a deal breaker but ideally, we would only push commands that changed.

**Comment by suneettipirneni** [comment]
> One small issue that I see is that this PR doesn't seem to push on a per command basis. Either it detects no changes, so nothing gets pushed, or it detects a change, and then everything gets pushed. Not necessarily a deal breaker but ideally, we would only push commands that changed.

True but as it stands now it's a much more efficient development improvement than the system we have now. We can add change per-command later if need be.

**Comment by rob-3** [comment]
Fair enough. If you address my 2 comments above I will approve this.

**Comment by rob-3** [comment]
I also might add per command pushing if you don't mind. I think it would be a quick addition with fairly significant value.

**Comment by suneettipirneni** [comment]
> I also might add per command pushing if you don't mind. I think it would be a quick addition with fairly significant value.

I'd prefer to get this merged in, as soon as possible because its needed for development. I'm happy to look at your changes in another PR.

**Comment by rob-3** [comment]
> > I also might add per command pushing if you don't mind. I think it would be a quick addition with fairly significant value.
> 
> I'd prefer to get this merged in, as soon as possible because its needed for development. I'm happy to look at your changes in another PR.

Sure thing. Only blockers for me are the question about `isEqual()` and suggestion to merge lines 73 and 74.

**Comment by rob-3** [comment]
Does `isEqual` not also cover the other checks as well?

**Comment by rob-3** [comment]
Like couldn't we just entirely replace `commandEquals` with `isEqual`?

**Comment by suneettipirneni** [comment]
> > > I also might add per command pushing if you don't mind. I think it would be a quick addition with fairly significant value.
> > 
> > 
> > I'd prefer to get this merged in, as soon as possible because its needed for development. I'm happy to look at your changes in another PR.
> 
> Sure thing. Only blockers for me are the question about `isEqual()` and suggestion to merge lines 73 and 74.

Done.

**Review by rob-3 — APPROVED** [review]

**Pull request merged**